[Lai Mei Tin] iP

1001mei opened the pull request:
## Duke
> "A journey of a thousand miles begins with a single step." - Chinese Proverb

Duke frees you from the hassle of remembering all the tasks you need to do. It's
- text-based
- fast to learn
- ~~EASY~~ *SUPER EASY* to use

All you need to do is
1. download it from [here](https://github.com/1001mei/ip/releases)
2. run `java -jar duke.jar` command in your terminal to launch the application
3. add your tasks
4. let it keeps track of the tasks for you :grin:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Sort tasks based on deadline (coming soon) 
- [x] Find tasks using keyword(s)

## Extra
If you are a Java programmer, feel free to tweak the code and customize it to better suits your needs!
Here is a snippet from the `main` method.
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

1001mei committed:
Add greet and exit functionality

1001mei committed:
Add echo functionality

1001mei committed:
Implement add task functionality

1001mei committed:
Implement list task functionality

1001mei committed:
Add a class to represent task

1001mei committed:
Add mark and unmark functionalities

1001mei committed:
Add three classes to represent types of tasks

1001mei committed:
Add support for tracking three types of tasks

1001mei committed:
Add a sample test for Duke

1001mei committed:
Add some error handling mechanisms

1001mei committed:
Replace array with ArrayList for storing tasks

1001mei committed:
Add support for deleting tasks from the list

1001mei committed:
Add support to store tasks in disk

1001mei committed:
Add ability to recognize dates in deadline tasks

1001mei committed:
Merge branch 'branch-Level-7'

1001mei committed:
Merge branch 'branch-Level-8'

1001mei committed:
Add Ui class

1001mei committed:
Add Storage class

1001mei committed:
Add TaskList class

1001mei committed:
Add Parser class

1001mei committed:
Add Command class and various command subclasses

1001mei committed:
Refactor Duke class to make it more OOP

1001mei committed:
Update .gitignore file

1001mei committed:
Organize classes into java packages

1001mei committed:
Add some JUnit tests

1001mei committed:
Update .gitignore file

1001mei committed:
Add some java documentations

1001mei committed:
Add find task functionality

1001mei committed:
Tweak the code to comply with coding standard

1001mei committed:
Merge branch 'branch-A-JavaDoc'

1001mei committed:
Merge branch 'branch-A-CodingStandard'

1001mei committed:
Merge branch 'branch-Level-9'

1001mei committed:
Add functionality to clear all tasks at once

1001mei committed:
Merge remote-tracking branch 'origin/add-gradle-support'

janald99 commented on src/main/java/duke/command/AddDeadlineTaskCommand.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Represents a command that when executed adds a deadline task to the task list.
+ */
+public class AddDeadlineTaskCommand extends Command {
```
I like how the names of your classes are super clear and straightforward 👍

lkldev replied:
Agree with @/janald99, great class names and easy to understand! 👍

1001mei committed:
Automate project builds using Gradle

janald99 commented on src/main/java/duke/command/AddDeadlineTaskCommand.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    public AddDeadlineTaskCommand(String description, LocalDate date) {
+        this.description = description;
+        this.date = date;
```
Perhaps the Bot could integrate time into Deadline as well?

1001mei replied:
I agree with you. Will implement this in the future.

janald99 commented on src/main/java/duke/command/AddDeadlineTaskCommand.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    /**
+     * Invokes <code>addDeadlineTask</code> method of <code>taskList</code> to add a deadline task.
```
I like your java doc comments a lot. They are very detailed 💯

janald99 commented on src/main/java/duke/command/AddEventTaskCommand.java:
```diff
@@ -0,0 +1,42 @@
+        taskList.addEventTask(description, at);
+
+        int taskIndex = taskList.getNumberOfTasks() - 1;
+        ui.showMessage("Got it. I've added this task: ");
```
What if there is a Messages class that consists of the messages that you want to show?
Since "Got it. I've added this task: " may be printed multiple times, maybe you can abstract that into a class and call ui.showMessage(Messages.TASK_ADDED_MSG).
Just an alternative to consider! 😄 It may or may not be more of a hassle.

1001mei replied:
This is a good idea! May add this class in the future.

janald99 commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,37 @@
+     */
+    @Override
+    public boolean isBye() {
+        return true;
```
an alternative(that i saw from someone else's code) you can consider is:
```
return this instanceof ByeCommand;
```

in the abstract Command class.
That may help readers understand better, but it could be less efficient.
What do you think about this?

1001mei replied:
I think the alternative is fine also, but I would still prefer overriding it in ByeCommand class because Command class knows nothing about its subclass, and I personally feel that writing code in superclass that refers to subclass is a bit weird. Just personal preferences I guess 😀

janald99 replied:
that makes a very good point. I am convinced this is better :)

janald99 commented on src/main/java/duke/command/ListTaskCommand.java:
```diff
@@ -0,0 +1,32 @@
+
+        ui.showMessage("Here are the tasks in your list:");
+        for (int i = 0; i < numberOfTasks; i++) {
+            ui.showMessage(String.format("%d. %s", i + 1, taskList.getDescriptionOfTaskAtIndex(i)));
```
perhaps the name of the method ```getDescriptionOfTaskAtIndex(i)``` is a little long? Although it explains very clearly what the method does. 🤔

janald99 reviewed (commented):
LGTM. The names of all variables and methods are super specific and detailed! The only trade-off for detailed names are that it may be too long, but it is worth it for readers to understand your code better, i believe 👍

1001mei committed:
Resolve all coding style violations

1001mei committed:
Set up the skeleton code for GUI

1001mei committed:
Add a new Main class

1001mei committed:
Organize GUI related classes into package

1001mei committed:
Adapt the code to use GUI

1001mei committed:
Add small modification to the GUI font

1001mei committed:
Merge branch 'branch-Level-10'

lkldev commented on src/main/java/duke/command/SetupCommand.java:
```diff
@@ -0,0 +1,26 @@
+/**
+ * Represents a command that when executed prepares the application for getting commands from user.
+ */
+public class SetupCommand extends Command {
```
I like how you have a SetupCommand to load the data from storage 👍 It makes the code neater without having the main to initialize the storage before running.

lkldev commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,71 @@
+ * This class acts as a parser that parses user inputs into specific commands.
+ */
+public class Parser {
+    private static final int OFFSET = 1;
```
Great use of constants to replace magic numbers 👍 However, I feel that maybe `ARRAY_INDEX_OFFSET` might be a more meaningful name?

1001mei replied:
I agree, thanks for the suggestion!

lkldev commented on src/test/java/duke/parser/ParserTest.java:
```diff
@@ -0,0 +1,50 @@
+
+import duke.exception.DukeException;
+
+public class ParserTest {
```
I like how you break all the test cases into different methods 👍 I would have created one method and tested everything related to the feature under testing in it. E.g. `parse_invalidMarkCommand` to test both `parse_invalidMarkCommandFormat` and `parse_incompleteMarkCommand`. But your way is definitely more comprehensive 👍

lkldev commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,176 @@
+        try {
+            numberOfTasks = data.length;
+            for (int i = 0; i < numberOfTasks; i++) {
+                String tokens[] = data[i].split(",");
```
Based on the java coding standard, the array specifiers must be attached to the type not the variable, e.g. 
```suggestion
                String[] tokens = data[i].split(",");
```

1001mei replied:
Thanks for pointing out this coding standard violation, will resolve it in coming commits.

lkldev commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,48 @@
+package duke.ui;
+
+import java.io.PrintStream;
+
```
Seems like there is an extra newline here between the same `java.*` import group.

lkldev commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.Scanner;
+
+public class Ui {
```
Based on the java coding standard, it will be good if you add comments for all public classes and methods.

lkldev reviewed (commented):
Overall, Great work! Love the descriptive method names you provided. Other than that, below are some nits I have picked out using the help of checkstyle, you may want to integrate it into your project to help you spot the coding standard related issues 👍

1001mei committed:
Add some assert statements

There are no assertions in the code.

Some important assumptions about the code cannot be verified.

Let's use the assert feature to document important assumptions about
internal invariants as well as preconditions and postconditions.

Adding assert statements increases the robustness of the code because
any violation of assumptions about the code will be detected.

1001mei committed:
TaskList class: make populateWith method more readable

PopulateWith method uses magic number to index the array in several
places.

Frequent access of the array using magic numbers makes the code
more clutter and less readable.

Let's extract the few elements in the array to meaningful named
variables, then replaced various array indexing with these variables.

Using variables with self-explanatory names is preferable than
indexing the same position of the array multiple times, because it
is hard to remember what each element in the array represents.

1001mei committed:
Merge pull request #1 from 1001mei/branch-A-Assertions

Add some assert statements

1001mei committed:
Merge branch 'master' into branch-A-CodeQuality

1001mei committed:
Merge pull request #2 from 1001mei/branch-A-CodeQuality

TaskList class: make populateWith method more readable

1001mei committed:
Add .yml file used for Continuous Integration

1001mei committed:
Move gradle.yml to correct directory

1001mei committed:
Fix minor checkstyle error

1001mei committed:
Make the command syntax more flexible

1001mei committed:
Merge branch 'branch-C-FriendlierSyntax'

1001mei committed:
Refactor the parse method in Parser.java

1001mei committed:
Fix method header comments format

1001mei committed:
Improve the GUI

1001mei committed:
Merge branch 'branch-A-BetterGui'

1001mei committed:
Fix minor checkstyle error

1001mei committed:
Add a user guide

1001mei committed:
Merge pull request #4 from 1001mei/branch-A-UserGuide

Add a user guide

1001mei committed:
Set theme jekyll-theme-architect

1001mei committed:
Set theme jekyll-theme-modernist

1001mei committed:
Update docs/README.md

1001mei committed:
Set theme jekyll-theme-cayman
